compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -238,18 +238,38 @@ data ConversionEnv = ConversionEnv
   , convRange :: !(Maybe SourceLoc)
   }
 
-newtype ConvertM a = ConvertM (ReaderT ConversionEnv (Except FileDiagnostic) a)
-  deriving (Functor, Applicative, Monad, MonadError FileDiagnostic, MonadReader ConversionEnv)
+data ConversionState = ConversionState
+    { freshTmVarCounter :: Int
+    , freshTyVarCounter :: Int
+    }
+
+newtype ConvertM a = ConvertM (ReaderT ConversionEnv (StateT ConversionState (Except FileDiagnostic)) a)
+  deriving (Functor, Applicative, Monad, MonadError FileDiagnostic, MonadState ConversionState, MonadReader ConversionEnv)
 
 instance MonadFail ConvertM where
     fail = conversionError
 
 runConvertM :: ConversionEnv -> ConvertM a -> Either FileDiagnostic a
-runConvertM s (ConvertM a) = runExcept (runReaderT a s)
+runConvertM s (ConvertM a) = runExcept (evalStateT (runReaderT a s) st0)
+  where
+    st0 = ConversionState
+        { freshTmVarCounter = 0
+        , freshTyVarCounter = 0
+        }
```

Comment:
Should you also reset the counters before each declaration, to prevent the order of declarations from affecting each other?

Reply:
I thought about that as well. My conclusion was that it's not important for deterministic compilation of the module since a changed order would change the result anyway. However, after thinking about it again, I realized that this sort of intra-definition determinism is desirable for testing the code generation of individual function. I'll reset the counter before each top-level declaration and each template.

Reply:
👍  Thanks!